### ISSUE TITLE
twisted.test.test_sip fails on Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8669]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)| @rodrigc reported|
|-|-|
|Trac ID|trac#8669|
|Type|enhancement|
|Created|2016-07-25 06:21:56Z|

If I do:

```
python3 bin/trial twisted.test.test_sip
```

I get:

```
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 520, in testCantForwardRequest
    r.addHeader("via", sip.Via("1.2.3.4").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testCantForwardRequest
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 513, in testLoop
    r.addHeader("via", sip.Via("1.2.3.4").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testLoop
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 460, in testReceivedRequestForward
    r.addHeader("via", sip.Via("1.2.3.4").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testReceivedRequestForward
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 491, in testReceivedResponseForward
    r.addHeader("via", sip.Via("127.0.0.1").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testReceivedResponseForward
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 440, in testRequestForward
    r.addHeader("via", sip.Via("1.2.3.4").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testRequestForward
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 480, in testResponseForward
    r.addHeader("via", sip.Via("127.0.0.1").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testResponseForward
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 502, in testResponseToUs
    r.addHeader("via", sip.Via("127.0.0.1").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testResponseToUs
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 474, in testResponseWrongVia
    r.addHeader("via", sip.Via("foo.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ProxyTests.testResponseWrongVia
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 604, in testFailedAuthentication
    self.register()
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 551, in register
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testFailedAuthentication
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 635, in testNoContactLookup
    self.register()
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 551, in register
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testNoContactLookup
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 563, in testRegister
    self.register()
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 551, in register
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testRegister
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 582, in testUnregister
    self.register()
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 551, in register
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testUnregister
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 628, in testWrongDomainLookup
    self.register()
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 551, in register
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testWrongDomainLookup
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 615, in testWrongDomainRegister
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testWrongDomainRegister
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 623, in testWrongToDomainRegister
    r.addHeader("via", sip.Via("client.com").toString())
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.RegistrationTests.testWrongToDomainRegister
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 364, in testNAT
    self.assertNotEqual(v.toString().find("rport=12345"), -1)
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.testNAT
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 307, in testSimpler
    self.checkRoundtrip(v)
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 253, in checkRoundtrip
    s = v.toString()
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.testSimpler
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 280, in test_complex
    self.assertEqual(v.toString(),
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_complex
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 327, in test_deprecatedRPort
    self.assertEqual(v.toString(), "SIP/2.0/UDP foo.bar:5060;rport")
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_deprecatedRPort
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 338, in test_rport
    self.assertEqual(v.toString(), "SIP/2.0/UDP foo.bar:5060;rport")
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_rport
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 349, in test_rportValue
    self.assertEqual(v.toString(), "SIP/2.0/UDP foo.bar:5060;rport=1")
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_rportValue
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 301, in test_simple
    self.assertEqual(v.toString(),
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_simple
===============================================================================
[ERROR]
Traceback (most recent call last):
  File "/Users/crodrigues/twisted_win/twisted/test/test_sip.py", line 373, in test_unknownParams
    self.assertEqual(v.toString(), s)
  File "/Users/crodrigues/twisted_win/twisted/protocols/sip.py", line 244, in toString
    etc.sort()
builtins.AttributeError: 'dict_items' object has no attribute 'sort'

twisted.test.test_sip.ViaTests.test_unknownParams
-------------------------------------------------------------------------------
Ran 46 tests in 0.282s

FAILED (failures=14, errors=1143, successes=7)
```

<details><summary>Searchable metadata</summary>

```
trac-id__8669 8669
type__enhancement enhancement
reporter__rodrigc rodrigc
priority__normal normal
milestone__Python_3_x Python-3.x
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1469427716126394 1469427716126394
changetime__1478019508574299 1478019508574299
version__None None
owner__rodrigc rodrigc

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_1"></a>|@rodrigc commented|
|-|-|

[#8669](https://github.com/twisted/twisted/issues/8669)

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

I'm guessing you meant to link to https://github.com/twisted/twisted/pull/580 :)

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)<a name="note_3"></a>|@moshez set owner to @rodrigc|
|-|-|

Reviewed on GH.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_4"></a>|@rodrigc set status to `closed`|
|-|-|

In changeset 361fb29b9661b1a3df5ac569127608ec84a7bd3f
```
#!CommitTicketReference repository="" revision="361fb29b9661b1a3df5ac569127608ec84a7bd3f"
Merge pull request #580 from twisted/8669-rodrigc-sip

Author: rodrigc
Reviewer: moshez
Fixes: #8669

Port twisted.protocols.sip to Python 3
```
